### PR TITLE
Plugins system for models and crop generators

### PR DESCRIPTION
**Pull request by santient**
These are some significant changes, so make sure I didn't mess anything up :P

### CODE REVIEW

**Comment by santient** [comment]
Another option is to have a python file as a parameter rather than a plugin system

**Comment by santient** [comment]
I'm reworking this a bit before it is merged

File: plugins/models/cnn.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import keras
+from keras.layers import *
+from keras.models import Model, Sequential
+from keras.optimizers import Adam
+
+from deepprofiler.learning.model import DeepProfilerModel
+
+
+def define_model(config, dset):
```
Comment (jccaicedo) [inline plugins/models/cnn.py]: Is this the VGG architecture? Add some comments to the class definition and also a link to the paper if possible. A short version of this example: https://github.com/tensorflow/tensorflow/blob/4665633c5f3fcd98de8c4cab909c8ac64c28ed0f/tensorflow/contrib/slim/python/slim/nets/vgg.py#L15

File: plugins/models/resnet18.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import keras_resnet
+import keras_resnet.models
+import tensorflow as tf
+
+from deepprofiler.learning.model import DeepProfilerModel
+
+
+def define_model(config, dset):
+
+    def make_regularizer(transforms, reg_lambda):
```
Comment (jccaicedo) [inline plugins/models/resnet18.py]: The regularizer can be dropped. Can you remove this function? We will not support this type of regularization going forward.

File: plugins/models/resnet18.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import keras
+import keras_resnet
+import keras_resnet.models
+import tensorflow as tf
+
+from deepprofiler.learning.model import DeepProfilerModel
+
```
Comment (jccaicedo) [inline plugins/models/resnet18.py]: We can add a link to the ResNet paper: https://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/He_Deep_Residual_Learning_CVPR_2016_paper.pdf

File: plugins/models/autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import keras
+from keras.layers import *
+from keras.models import Model, Sequential
+from keras.optimizers import Adam
+
+from deepprofiler.learning.model import DeepProfilerModel
+
+
```
Comment (jccaicedo) [inline plugins/models/autoencoder.py]: Great job implementing this autoencoder. We need to find a paper to link here too. Also, although the code is pretty much self-explanatory, try adding some comments to structure code in blocks that humans can read.

File: plugins/models/vae.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import keras
+from keras import backend as K
+from keras import objectives
+from keras.layers import *
+from keras.models import Model, Sequential
+from keras.optimizers import Adam
+
+from deepprofiler.learning.model import DeepProfilerModel
```
Comment (jccaicedo) [inline plugins/models/vae.py]: Excellent job with the VAE! Same comments here. Let's try to add a link to a paper, and also add some basic comments.

**Comment by jccaicedo** [comment]
This branch is exciting! The refactoring is quite good, and we have new models in to try using a plugins strategy.

@santient have you run a training loop with the small development set to make sure all the models in this branch run without errors?

**Comment by santient** [comment]
Should this pr be replaced with the newer one integrating this with the other changes

**Pull request merged**